### DISCUSSION TITLE
Block Editor

### DISCUSSION THREAD

**Discussion by tarwin** [directus/directus#5776] [Feature Requests]
It would be great to add a **Block Editor** aka _Medium_ or the new _Wordpress_.

The main usefulness of this versus a WYSIWYG editor is it can be controlled in a better way with only very specific output, which is then easier to transform with CSS / JS when in use.

**Comment by benhaynes** [comment]
I agree. I think a simple static JSON version of this could be added to Core, and the Many-to-Any Interface can be used for more "edge-case" uses. No reason to reinvent the wheel here, so I think the best thing to do is to create a new interface that leverages: https://github.com/codex-team/editor.js

Yeehaw?

**Reply by rijkvanzanten** [reply]
🤠

**Reply by tarwin** [reply]
Sounds great! If it saves JSON would you need a special converter (sorry just started so not sure about internal lingo) be needed for `field/list view`?

What would the M2M interface be used for?

**Reply by benhaynes** [reply]
I don't think we would want an M2M or M2A relational block editor in Core... it would be a configuration nightmare trying to make that work for any/all project needs. I think JSON is the way to go. We'd likely want a dedicated Display to show this specific JSON, since it might be able to show something more useful than the typical JSON _code_ display.

So the Interface would be a full-featured and customizable block editor and the Display would likely be some custom inline block readout for the JSON.

**Reply by FullStackAlex** [reply]
What about GrapesJS?:
https://github.com/artf/grapesjs

**Reply by FullStackAlex** [reply]
It‘s the most Webflow like open source builder and all the no code folks are all crazy about Webflow.

**Reply by tarwin** [reply]
GrapeJS looks great but I don't think it is useful as a field component. It seems more for building full pages, rather than something that can be used to here. Happy to be convinced otherwise.

**Reply by FullStackAlex** [reply]
@tarwin It would be just a builder for the content field. Nothing more. Why shouldn't the content field not contain html markup as a string? The only problem might be related javascript and css code (if some fancy component/template is used for building the content). This how all the page builders (Elementor, Brizzy, Visual Composer etc) work within WordPress. They just replace the post_content field and "enqueue" some additional css and js files.

**Reply by tarwin** [reply]
Understand why some people would want a more powerful editor, and for this GrapeJS could make sense. But in my experience you get into a lot of trouble when you allow people to edit everything.

The idea of a block editor is to allow non-super users to do things like add blog posts. These can then be styled as required by whatever frontend you are using, while also guaranteeing to not break your site.

Again, not against the idea. It's just got different use cases than a block editor.

**Reply by benhaynes** [reply]
Even if this doesn't make it into Core, I think this granular editor could be a good interface extension... or even a dedicated Module for editing full pages. Also, it might be useful to have when we allow editing Email Templates (for our transactional emails). Then admins could design/build/edit their email templates in the App. 🤔

**Reply by FullStackAlex** [reply]
@tarwin 
Ah, ok, see your point. Then maybe two different field types like "page-builder" and "block-editor" 🙃

**Reply by tarwin** [reply]
Yup, you've got it. 😄

**Reply by paescuj** [reply]
> I agree. I think a simple static JSON version of this could be added to Core, and the Many-to-Any Interface can be used for more "edge-case" uses. No reason to reinvent the wheel here, so I think the best thing to do is to create a new interface that leverages: https://github.com/codex-team/editor.js
> 
> Yeehaw?

I kinda like [Keystone's implementation](https://keystonejs.com/docs/guides/document-fields) of a "Block editor".
Content is stored as JSON and editor is based on [Slate](https://docs.slatejs.org/).

**Reply by tylerforesthauser** [reply]
If we're just pointing out editors that store their content as JSON-structured data and could be used as block editors, than I'd throw in [Tiptap](https://tiptap.dev/installation/vue3) (based on [Prosemirror](https://prosemirror.net/)) as it is framework-agnostic. It could be extended to include page-builder type nodes to accommodate that usage but could otherwise be an alternate WYSIWYG editor.

**Reply by kytosai** [reply]
I really agree with the choice of editor.js for the block editor implementation

A real demonstration of editing with a block editor like the editor.js library is the notion app. Although the two are not exactly the same, the writing and usage are very similar

I have experienced editor.js for 1 year and notion for 3 years.

Recently when notion has an api, I checked the response of notion content and found it to be much more detailed and flexible than the editor.js library.

**Reply by frontendmiczone** [reply]
I highly recommend using editor.js and saving the data as json.

I have used this package https://github.com/dimitrov-adrian/directus-extension-editorjs-interface . However, it is not optimized well enough, there are still many errors.

Storing json data will make it much easier and more flexible to handle in the frontend

**Reply by lukas-schaetzle** [reply]
+1 for Slate.js. Editor.js looks also nice but it doesn't support nested blocks (there's a plugin for it but it works kinda meh...). With Slate you can do pretty much anything you want as far as I can tell.

**Reply by CanRau** [reply]
Might be a bit new and the community accordingly small but wanted to throw lexical.dev in the mix, from my current experiments I like it a lot so far

**Reply by tarwin** [reply]
On nested blocks. You may want this at some point but I'd suggest keeping things as simple as possible (no nesting) to start with is a good idea. This way when you want to parse the JSON (to HTML in an SSR or frontend) things are simple.

**Comment by thomasazar** [comment]
It may be worth mentioning that a great deal of effort has been put into an interface for editor.js over at [this repo](https://github.com/dimitrov-adrian/directus-extension-editorjs-interface). I've been using it and playing around with it for a few months, and it does what it says on the tin: everything is stored as JSON.

Any chance it could be pulled into the core app?

**Reply by benhaynes** [reply]
I am down to get this reviewed, yes. I definitely like the idea of saving as JSON so the data is more agnostic.

To make it into Core, we'll need to look at the actual implementation to ensure the code is aligned with our standards. Also, we'll need to find time to review... and the team is a bit busy preparing for Directus 9 at the moment. 🙂

**Reply by thomasazar** [reply]
Of course the release of Directus 9 is the priority! I'm eagerly looking forward to it, and y'all have been pouring in the work. Thank you for everything the team has done.

I want to make it clear (just in case it wasn't obvious) that I am not in any way affiliated with the editor.js interface, other than being a user of it. I don't want to speak on behalf of @dimitrov-adrian, who is the brains behind the interface.

**Comment by ghost** [comment]
For me a block editor would be a rich text editor, that saves JSON code and allows to insert relations. I don’t think that it should be a M2A, I think it should have the ability to hold multiple M2A, M2M, etc.. as blocks. Besides that it should also offer the ability to insert custom inline marks, with custom fields, that output JSON too.

Use case: Let’s say I want to insert multiple Gallery blocks into my Block Editor field. Each Gallery can contain Images or Videos. And maybe I also want to insert other blocks, like Persons … and so on. This can’t be realized with if the Block Editor is a M2A. I think it must be a M2A field, where it can reference an other M2A collection — I don’t know if that’s possible in the current version of directus. A simple solution could be to make an own block editor collection, where we add O2M, M2M, M2A or other fields, and then reference this collection as our block editor in our main collection.

I did couple of projects with sanity.io and I would highly recommend you to check out their block editor (they call it “portable text”). The strength of the editor is to insert anything you can think of: inline as well as blocks. And the output is always JSON. On the frontend when retrieving the data from the API you can render that data like you wish — no boundaries.

I think for directus we can achieve this with [Editor.js](https://editorjs.io/). The only (huge) drawback with Editor.js is, that we can’t store inline marks as JSON :( — but they work on it as you can [see on their roadmap](https://github.com/codex-team/editor.js#roadmap). I’ve also looked at [tip tap](https://tiptap.dev/) which looks like a very promising editor, but it seems like the project isn't moving forward that fast. Then I looked at [quilljs](https://quilljs.com/) and [slatejs](https://www.slatejs.org/), but somehow I feel more comfortable with [Editor.js](https://editorjs.io/).

Imagine what could be possible. This would be amazing!

**Reply by CanRau** [reply]
Not sure if has been mentioned but for an RTE is lexical.dev

**Comment by madc** [comment]
By now we are running multiple Directus installations mostly as headless CMS [with EditorJS](https://github.com/dimitrov-adrian/directus-extension-editorjs-interface) and it is really awesome to work with.
As it formats the content as an array of JSON objects instead of a block of HTML, it moves most of the rendering power to the frontend code.

**Reply by rijkvanzanten** [reply]
How (if at all) are you dealing with relationships? The major downside that I've been trying to solve for is that you can no longer rely on nested relational data fetching from the API as the nested JSON is just a blob rather than columns with relationships 🤔

**Reply by madc** [reply]
Not sure, if I'm understanding you correctly. Please feel free to correct me.

We've two cases where relations are involved: images and attachments (files).
In both cases there is no real relation but rather a one-sided object containing id son some other information. Whe then use this to render the image or a link to the attachment in frontend. If the images/files are changed or deleted, its not reflected.
It would be cool to have a better solution, but for now I can live with it. And it is way more flexible than having a block of HTML.

**Reply by madc** [reply]
Just a quick thought: Currently the EditorJs extension stores the JSON-Object in a single text column. Maybe there is a way of implementing it similar to a repeater or M2A interface.

**Reply by LasseRosenow** [reply]
>Maybe there is a way of implementing it similar to a repeater or M2A interface.

Not sure if this is in scope of this thread, but using editor.js not only like, but also "AS" an alternative interface for M2A relations would be really cool, so one could edit the M2A entries inline instead of having to open them all the time :)

**Comment by phihu** [comment]
I understand the complexity of this (potential) feature, but having this would be a huge deal (this and version control are possibly the main issues we encounter with our clients/projects).
 
Could it be a starting point, to use editor.js and create a custom editor JS block, that basically shows the interface selecting/creating related elements?
The field data (in case of a newly created element) or the related key (for an existing element) would have to be stored in the editor.js - JSON and handled/transformed when saving.

The display of a related element within the editor could be done according to the element's/collection's display settings as a simple list (for now). A next step would be a way of creating custom blocks for editing/displaying an element.

**Reply by madc** [reply]
I'm all for having editor.js officially supported, but meanwhile have you had a look at this plugin?
https://github.com/dimitrov-adrian/directus-extension-editorjs-interface

We are using it quite extensively and it works pretty well!

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by rijkvanzanten** [comment]
This was since added! #18525
